Param: add locking

bkueng opened the pull request:
This adds locking to the param module.

We need to protect access to the param_values array. This is dynamically
allocated and resized (utarray_reserve() calls realloc). If some thread
was iterating the array while another was resizing the array, the first one
would iterate on a freed array, thus accessing invalid memory.

On NuttX this could lead to hardfaults in rare conditions.

@davids5 This also adds locking to the flash-based params. Can you please test that I called `param_init` everywhere where needed? I checked where `hrt_init` was called and compared that against the cmake config's that enable the `param` module.

@Stifael can you please test this on Snappy?

I tested on Posix & Pixracer.

This could be implemented more efficiently if we had atomic variables.

davids5 commented on src/modules/systemlib/param/param.c:
```diff
@@ -127,6 +127,7 @@ get_param_info_count(void)
 {
 	/* Singleton creation of and array of bits to track changed values */
 	if (!param_changed_storage) {
+		/* Note that we have a (highly unlikely) race condition here: in the worst case the allocation is done twice */
```
@bkueng 

Would you be able to avoid the param_init() and uses this singleton to init the semaphore.  Also wrap this in a schedule_lock to avoid the race

bkueng replied:
I thought about this, but that's really a chicken-egg problem: adding a singleton requires synchronization itself to be thread-safe.
And static global object initializers should be avoided (Not sure but I thought they are not supported on NuttX)

I'll look into it but sched_lock is really only useful for NuttX.

davids5 commented:
Any hope of adressing the rest of this https://github.com/PX4/Firmware/issues/6556 ?

bkueng commented:
> The flash is written to for every change. A fresh boot after "param reset" results in X * number of param writes to flash. When the param save command was needed before this would be 1 write operation.

I noticed that too. This considerably slows down boot after `param reset`. We could add a callback after X seconds to store changed params. But I don't have time to implement and test this at the moment.

LorenzMeier commented:
> The flash is written to for every change. A fresh boot after "param reset" results in X * number of param writes to flash. When the param save command was needed before this would be 1 write operation.

That's actually a bug in param_reset() then. It should call the API which doesn't do the autosave and save at the end. The autosave API should only be used for single-param operations like `param set` on the shell.

LorenzMeier commented:
@bkueng There seems to be a deadlock in the param test, see the Circle CI output:

```
Test project /home/ubuntu/Firmware/build_unittest

    Start 1: param_test


command ./Tools/docker_run.sh 'make tests_coverage' took more than 10 minutes since last output
```

davids5 commented:
> That's actually a bug in param_reset() then. It should call the API which doesn't do the autosave and save at the end. The autosave API should only be used for single-param operations like param set on the shell.

What is happening is that the reset dumps the array. then every shell operation ```param set ...``` does a write. It use to be bounded and a ```param save``` was needed. We lost that along the way

LorenzMeier commented:
The old param save approach created a vulnerability for developers to wipe out their board. That happened like 20 times and PX4 was considered flaky (because they commonly typed "param save", then cycled power).

LorenzMeier commented:
@bkueng You can repro the hang with `make tests` which will hang in the param test call.

dagar commented:
@bkueng If it's anything but a trivial fix you might want to consider killing off that standalone param test and moving it within the tests command using real parameters. https://github.com/PX4/Firmware/blob/master/unittests/param_test.cpp

bkueng commented:
@dagar I fixed the test, I needed to add `param_init` to the setup call.

Additionally I made the `param show` command case insensitive (eg. `param show sys_*` works) and added the option `-c` to show only changed params.

For the multiple saving issue: we could rate limit the saving. That would solve both problems. The first `param set` would save immediately, then the next would save at least 1 second (or so) later.

LorenzMeier commented:
Rate limiting doesn't sound terrible at all.
